File: Go/Day09.go
package main

import Helper "./adventhelper"

func test(arr []int, offset int, size int) (bool) {
	if len(arr) < offset + size + 1 {
		return false
	}

	for i := 0; i < size; i++ {
		for j := i + 1; j < size; j++ {
			if arr[offset+i] + arr[offset+j] == arr[offset + size] {
				return true
			}
		}		
	}

	return false
}

func findRange(arr []int, target int, blocksize int) ([]int) {
	block := arr[:blocksize]
	s := 0
	for _, i := range block {
		s += i
	}

	if s == target {
		return block
	} else if s > target {
		return findRange(arr[1:], target, 2)
	} else {
		return findRange(arr, target, blocksize+1)
	}
}

func main() {
	is, _ := Helper.ReadIntFile("../input/input09.txt")
	size := 117

	var part1 int
	for i := 0; i < len(is) - size; i++ {
		if !test(is, i, size) {
			part1 = is[i+size]
			break
		}
	}
	max, min := Helper.MaxAndMin(findRange(is, part1, 2))

	Helper.PrintSoln(9, part1, max + min)

	return
}

File: Haskell/Day09.hs
import AdventHelper

block :: Int -> [Integer] -> Int -> ([Integer], Integer)
block size is offset = (take size $ drop offset is,
                        head $ drop (offset+ size) is)

test :: ([Integer], Integer) -> Bool
test ([_], _) = False
test ((x:xs), n)
  | elem (n - x) xs = True
  | otherwise       = test (xs, n)

findRange :: [Integer] -> Integer -> Int -> [Integer]
findRange xs target n
  | s == target = xs'
  | s >  target = findRange (drop 1 xs) target 2
  | s <  target = findRange xs target (succ n)
  where xs' = take n xs
        s = sum xs'


main = do
  putStrLn "Day 9"
  f <- readFile "../input/input09.txt"
  let ns = map (read) $ lines f :: [Integer]
  let size = 117

  let q = head $ filter (not . test . block size ns) [0,1..]
  let r = findRange ns (ns!!(q+size)) 2

  printSoln 1 (ns!!(q+size))
  printSoln 2 (maximum r + minimum r)
